src/shared/pkcs11-util.c

```diff
@@ -508,22 +508,65 @@ int pkcs11_token_read_x509_certificate(
 }
 #endif
 
+static int pkcs11_token_find_private_key_from_template(
+        CK_FUNCTION_LIST *m,
+        CK_SESSION_HANDLE session,
+        CK_ATTRIBUTE_PTR template,
+        CK_ULONG templateCount,
+        CK_OBJECT_HANDLE_PTR objects,
+        CK_ULONG objectCount,
+        CK_ULONG *ret_nobjects) {
+
+        CK_RV rv, rv2;
+
+        assert(m);
+        assert(session);
+        assert(template);
+        assert(templateCount);
+        assert(objects);
+        assert(objectCount);
+        assert(ret_nobjects);
+
+        rv = m->C_FindObjectsInit(session, template, templateCount);
+        if (rv != CKR_OK)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO),
+                                       "Failed to initialize object find call: %s", p11_kit_strerror(rv));
+
+        rv = m->C_FindObjects(session, objects, objectCount, ret_nobjects);
+        rv2 = m->C_FindObjectsFinal(session);
+        if (rv2 != CKR_OK)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO),
+                                       "Failed to finalize object find call: %s", p11_kit_strerror(rv));
+        if (rv != CKR_OK)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO),
+                                       "Failed to find objects: %s", p11_kit_strerror(rv));
+
+        return 0;
+}
+
 int pkcs11_token_find_private_key(
                 CK_FUNCTION_LIST *m,
                 CK_SESSION_HANDLE session,
                 P11KitUri *search_uri,
-                CK_OBJECT_HANDLE *ret_object) {
+                CK_OBJECT_HANDLE *ret_object,
+                CK_KEY_TYPE *ret_key_type) {
 
-        bool found_decrypt = false, found_class = false, found_key_type = false;
+        bool found_decrypt = false, found_class = false, found_derive = false;
         _cleanup_free_ CK_ATTRIBUTE *attributes_buffer = NULL;
         CK_ULONG n_attributes, a, n_objects;
         CK_ATTRIBUTE *attributes = NULL;
         CK_OBJECT_HANDLE objects[2];
-        CK_RV rv, rv2;
+        CK_ATTRIBUTE key_type_template = {
+                .type = CKA_KEY_TYPE,
+        };
+        CK_RV rv;
+        const CK_BBOOL yes = true;
```

Comment:
const without static on a non-pointer is weird. I figure this should be static as well